File: 0005/5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        # Monacher's Algorithm
        sp = "#" + "#".join(list(s)) + "#"

        pr = [0 for i in sp]

        c = 0
        r = 0

        while c < len(sp):
            while c-(r+1) >= 0 and c+(r+1) < len(sp) and sp[c-(r+1)] == sp[c+(r+1)]:
                r += 1

            pr[c] = r
            
            co = c
            ro = r
            c += 1
            r = 0

            while c <= co + ro:
                mc = co - (c - co)
                mmr = co + ro - c
                if pr[mc] < mmr:
                    pr[c] = pr[mc]
                    c += 1
                   
                elif pr[mc] > mmr:
                    pr[c] = mmr
                    c += 1
                else:
                    r = mmr
                    break
        longest_r = max(pr)
        start = pr.index(longest_r) - longest_r
        end = pr.index(longest_r) + longest_r
        ret = "".join(sp[start:end+1].split("#"))
        return ret



print(Solution().longestPalindrome("mom"))
# print(Solution().longestPalindrome("cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"))

File: 0005/5.cs
using System;

namespace Solution5
{
    class Solution
    {
        static void Main(string[] args)
        {
            Console.WriteLine(LongestPalindrome("mom"));
        }

        public static string LongestPalindrome(string s)
        {
            // Monacher's Algorithm
            var sp = "#";
            foreach (var item in s)
            {
                sp += item + "#";
            }
            var pr = new int[2 * s.Length + 1];

            var c = 0;
            var r = 0;


            while (c < sp.Length)
            {
                while (c - (r + 1) >= 0 && c + (r + 1) < sp.Length && sp[c - (r + 1)] == sp[c + (r + 1)])
                {
                    r++;
                }

                pr[c] = r;

                var oc = c;
                var or = r;
                c++;
                r = 0;
                while (c <= oc + or)
                {
                    var mc = oc - (c - oc);
                    var mmr = oc + or - c;
                    if (pr[mc] < mmr)
                    {
                        pr[c] = pr[mc];
                        c++;
                    }
                    else if (pr[mc] > mmr)
                    {
                        pr[c] = mmr;
                        c++;
                    }
                    else
                    {
                        r = mmr;
                        break;
                    }
                }
            }
            var t = get_max(pr);
            var y = get_max_index(pr);
            var temp = sp.Substring(y-t,y+t+1);
            var longest_palindrome_in_S = "";
            foreach (var item in temp)
            {
                if(item != '#')
                {
                    longest_palindrome_in_S += item;
                }
            }
            return longest_palindrome_in_S;
        }

        private static int get_max_index(int[] pr)
        {
            int max = 0;
            int index = 0;
            for(int i = 0; i < pr.Length; i++)
            {
                var item = pr[i];
                if (item > max)
                {
                    max = item;
                    index = i;
                }
            }
            return index;
        }

        private static int get_max(int[] pr)
        {
            int max = 0;
            foreach (var item in pr)
            {
                if (item > max)
                    max = item;
            }
            return max;
        }
    }
}
